internal/controllers/service_controller.go

```diff
@@ -0,0 +1,147 @@
+)
+
+// ServiceReconciler reconciles a Pod object
+type ServiceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=pods/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=pods/finalizers,verbs=update
+func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("service-reconciler")
+
+	l.Info("Reconciling changed Service resource", "changed", req.NamespacedName)
+
+	svc := &corev1.Service{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, svc); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if val, ok := svc.Annotations["kusk-gateway/openapi-url"]; ok {
+		l.Info(`Detected "kusk-gateway/openapi-url" annotation`, "found", val)
+		openapi, err := getOpenAPIfromURL(val)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		var yml map[string]interface{}
+		err = yaml.Unmarshal(openapi, &yml)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		service := map[string]interface{}{"service": map[string]interface{}{
+			"name":      req.Name,
+			"namespace": req.Namespace,
+			"port":      svc.Spec.Ports[0].Port,
+		}}
+		upstream := map[string]interface{}{
+			"upstream": service,
+		}
+
+		if val, ok := yml["x-kusk"]; !ok {
+			l.Info(`"x-kusk" extension not found in OpenAPI definition`, "missing", val)
+			yml["x-kusk"] = upstream
+		} else if xkusk, ok := val.(map[string]interface{}); ok {
+			if _, contains := xkusk["upstream"]; !contains {
+				xkusk["upstream"] = service
+			}
+		}
+
+		yamlPayload, err := yaml.Marshal(yml)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		gatewaySpec := gateway.APISpec{
+			Spec: string(yamlPayload),
+		}
+
+		fmt.Println("payload", gatewaySpec.Spec)
+		api := &gateway.API{}
+
+		if err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, api); err != nil {
+			if errors.IsNotFound(err) {
+				//create
+				api.Name = req.Name
+				api.Namespace = req.Namespace
+				api.Spec = gatewaySpec
+				if err := r.Client.Create(ctx, api, &client.CreateOptions{}); err != nil {
+					l.Error(err, "error occured while creating API")
+					return ctrl.Result{}, err
+				}
+
+			}
+		} else {
+			api.Spec = gatewaySpec
+			if err := r.Client.Update(ctx, api, &client.UpdateOptions{}); err != nil {
+				l.Error(err, "error occured while updating API")
+				return ctrl.Result{}, err
+
+			}
+		}
+
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Service{}).
+		Complete(r)
+}
+
+func getOpenAPIfromURL(url string) ([]byte, error) {
```

Comment:
url is not guaranteed to be a valid url. We should parse it first and return an error if it's not good

Reply:
it doesn't matter if it isn't it will return error when it tries to get it